### ISSUE TITLE
miscompile with slicing test case

### ISSUE THREAD

**Issue by silvasean** [google/iree#8825]
**Describe the bug**

Miscompile of the following Torch code when invoked with `tensor<4xf32>`

```
    def forward(self, bbox1):
        return 3 * bbox1[2:]
```

**To Reproduce**

```
$ iree-run-mlir /tmp/repro.mlir -function-input="4xf32=0 1 2 3" -iree-hal-target-backends=dylib
EXEC @forward
result[0]: hal.buffer_view
2xf32=0 3
```

Expected result: `2xf32=6 9`

```
#map = affine_map<(d0) -> (d0)>
module attributes {torch.debug_module_name = "IouOfModule"} {
  func @forward(%arg0: tensor<?xf32>) -> tensor<?xf32> {
    %c0 = arith.constant 0 : index
    %c0_i64 = arith.constant 0 : i64
    %c3_i64 = arith.constant 3 : i64
    %c2_i64 = arith.constant 2 : i64
    %0 = tensor.dim %arg0, %c0 : tensor<?xf32>
    %1 = arith.index_cast %0 : index to i64
    %2 = arith.addi %c2_i64, %1 : i64
    %3 = arith.cmpi sge, %c2_i64, %c0_i64 : i64
    %4 = arith.select %3, %c2_i64, %2 : i64
    %5 = arith.cmpi slt, %4, %c0_i64 : i64
    %6 = arith.select %5, %c0_i64, %4 : i64
    %7 = arith.cmpi sgt, %6, %1 : i64
    %8 = arith.select %7, %1, %6 : i64
    %9 = arith.index_cast %8 : i64 to index
    %10 = arith.cmpi sge, %0, %9 : index
    %11 = arith.select %10, %0, %9 : index
    %12 = arith.subi %11, %9 : index
    %13 = tensor.extract_slice %arg0[%9] [%12] [1] : tensor<?xf32> to tensor<?xf32>
    %14 = linalg.init_tensor [%12] : tensor<?xf32>
    %15 = linalg.generic {indexing_maps = [#map, #map], iterator_types = ["parallel"]} ins(%13 : tensor<?xf32>) outs(%14 : tensor<?xf32>) {
    ^bb0(%arg1: f32, %arg2: f32):
      %16 = arith.sitofp %c3_i64 : i64 to f32
      %17 = arith.mulf %arg1, %16 : f32
      linalg.yield %17 : f32
    } -> tensor<?xf32>
    return %15 : tensor<?xf32>
  }
}
```

**Comment by silvasean** [comment]
It looks like vmvx doesn't have the bug.

**Comment by hanhanW** [comment]
Looks like a bug in cpu codegen, I'll take a look at it.

Does it happen in vulkan-spirv or cuda backends?

**Comment by silvasean** [comment]
I don't have vulkan-spirv or cuda hooked into my test suite yet.

**Comment by hanhanW** [comment]
vulkan backend is producing the wrong results.

```
iree-run-mlir -iree-hal-target-backends=vulkan-spirv -function-input="4xf32=0 1 2 3" ~/issue-8825.mlir
```

**Comment by MaheshRavishankar** [comment]
This might be fixed by #8793 . Trying

**Comment by MaheshRavishankar** [comment]
Yup thats the issue . With my change 

```
iree-run-mlir -function-input="4xf32=0 1 2 3" -iree-hal-target-backends=dylib repro_8825.mlir 
I /home/ravishankarm/Development/iree/iree/iree/tools/iree-run-mlir-main.cc:205] Compiling for target backend 'dylib*'...
I /home/ravishankarm/Development/iree/iree/iree/tools/iree-run-mlir-main.cc:308] Evaluating all functions in module for driver 'dylib'...
I /home/ravishankarm/Development/iree/iree/iree/tools/utils/vm_util.cc:154] Creating driver and device for 'dylib'...
EXEC @forward
result[0]: hal.buffer_view
2xf32=6 9
```
